docs/design/proposals/non-graceful-node-shutdown.md

```diff
@@ -0,0 +1,136 @@
+as unavailable _refer [k8s doc](https://kubernetes.io/docs/concepts/cluster-administration/node-shutdown/#non-graceful-node-shutdown)_. This results in the forceful deletion of pods
+scheduled onthe node and the cleanup of associated VolumeAttachment objects.
+However, this forceful cleanup sequence bypasses key CSI lifecycle calls most
+notably, the `NodeUnpublishVolume` and `NodeUnstageVolume` entry points are not
+invoked prior to `ControllerUnpublishVolume` call.
+
+## Problem
+
+When `ControllerUnpublishVolume` is called without the node first having cleaned
+up the volume (via `NodeUnstageVolume` and `NodeUnpublishVolume`), the CSI driver
+has no opportunity to revoke the node's access to the volume. The node may still
+hold active mounts, open file handles, or client sessions. This can lead to data
+corruption as applications may still be running on the broken node with active
+client sessions, even though the node is marked as out of service.
+
+## Proposed solution
+
+To ensure safe volume reuse and prevent stale client access during node disruptions,
+the proposed solution is to track the client address during the `NodeStageVolume()`
+operation and store it in the image or subvolume metadata under the key:
+`.csi.ceph.io/clientAddress/<NodeId>`.
+
+> Note: This metadata should not be copied when creating clones, snapshots,
+or mirror images of the volume.
+
+This stored address can then be used:
+
+- In `ControllerUnpublishVolume()` to blocklist or evict the client if the node has
+the `out-of-service` taints, ensuring it no longer has access to the volume.
+- In `ControllerPublishVolume()` to remove the client from the blocklist, allowing
+the volume to be accessed again when the node recovers.
+
+## Implementation Details
+
+### ControllerPublishVolume()
+
+- Check if the metadata key `.csi.ceph.io/clientAddress/<NodeId>`
+exists(`NodeId` from the request)
+- If it exists:
+  - Remove the client from the blocklist:
+
+    ```bash
+    ceph osd blocklist rm <clientAddress+nonce>
+    ```
+
+  - Remove the metadata key after removing from blocklist
+
+    ```
+    rbd image-meta remove <pool>/<image> csi.ceph.io/clientAddress/<NodeId>
+    ```
+  
+### NodeStageVolume()
+
+- **Applicable to**: RBD and CephFS
+- Before volume staging
+  - Retrieve the client address:
+
+  - Always set it in the image/subvolume metadata:
+    (`NodeId` from the plugin container argument `--nodeid`)
+
+    ```
+    .csi.ceph.io/clientAddress/<NodeId>: <clientAddress+nonce>
+    ```
+
+  - Continue with Volume staging
+
+### ControllerUnpublishVolume()
+
+- Retrieve the Node object using the provided `NodeId` from the request.
+- If the node has the `out-of-service` taint:
+  - Check for the metadata key:
+
+    ```
+    .csi.ceph.io/clientAddress/<NodeId>
+    ```
+
+  - If present, proceed to revoke client access:
+    - **For RBD**:
+
+      ```bash
+      ceph osd blocklist add <clientAddress+nonce>
```

Comment:
A "problem" with OSD blocklist entries is that by default they expire after 1 hour.  The expiration timeout is in seconds and can be changed by passing an additional integer argument.  For example, `ceph osd blocklist add 1.2.3.4/1234 86400` would persist the blocklist entry for 1 day.

If the goal here is to accommodate scenarios where the admin applies the out-of-service taint without knowing that the node is actually powered off, the blocklist entry needs to persist until it becomes known for sure that the node went through (an equivalent of) a power cycle.  Allowing the blocklist entry to expire before that point can lead to "stale client access" and therefore data corruption.